xwiki-pro-macros-api/src/main/java/com/xwiki/macros/viewfile/internal/ThumbnailGenerator.java

```diff
@@ -0,0 +1,258 @@
+     */
+    public byte[] getThumbnailData(AttachmentReference attachmentReference)
+    {
+        try {
+            File tempDir = new File(environment.getTemporaryDirectory(),
+                String.format(THUMBNAILS_PATH, attachmentReference.getDocumentReference().toString()));
+
+            File thumbnail = new File(tempDir, attachmentReference.getName() + JPG_EXTENSION);
+            if (!thumbnail.exists()) {
+                return generateAndGetThumbnailBytes(attachmentReference);
+            } else {
+                return Files.readAllBytes(thumbnail.toPath());
+            }
+        } catch (Exception e) {
+            logger.error("There was an error while attempting to get the thumbnail byte data. Root cause is: [{}]",
+                ExceptionUtils.getRootCauseMessage(e));
+            return new byte[0];
+        }
+    }
+
+    private byte[] generateAndGetThumbnailBytes(AttachmentReference attachmentReference) throws Exception
+    {
+        String extension = getExtension(attachmentReference.getName());
+        if (OFFICE_EXTENSIONS.contains(extension)) {
+            return getOfficeThumbnailBytes(attachmentReference);
+        } else if (PRESENTATION_EXTENSIONS.contains(extension)) {
+            return getPresentationThumbnailBytes(attachmentReference, extension);
+        } else if (extension.equals("pdf")) {
+            return getPDFThumbnailBytes(attachmentReference);
+        } else {
+            logger.warn("Extension type not supported.");
+            return new byte[0];
+        }
+    }
+
+    private byte[] getPDFThumbnailBytes(AttachmentReference attachmentReference) throws Exception
+    {
+        XWikiContext wikiContext = wikiContextProvider.get();
+        XWiki wiki = wikiContext.getWiki();
+        XWikiDocument document = wiki.getDocument(attachmentReference.getDocumentReference(), wikiContext);
+        InputStream is = document.getAttachment(attachmentReference.getName()).getContentInputStream(wikiContext);
+        return generateThumbnail(new ByteArrayInputStream(is.readAllBytes()), attachmentReference);
+    }
+
+    private byte[] getOfficeThumbnailBytes(AttachmentReference attachmentReference) throws Exception
+    {
+        ByteArrayInputStream bais = getPDFContent(attachmentReference);
+        return generateThumbnail(bais, attachmentReference);
+    }
+
+    private ByteArrayInputStream getPDFContent(AttachmentReference attachmentReference) throws Exception
+    {
+        XWikiContext wikiContext = wikiContextProvider.get();
+        XWiki wiki = wikiContext.getWiki();
+        XWikiDocument document = wiki.getDocument(attachmentReference.getDocumentReference(), wikiContext);
+        try (InputStream is = document.getAttachment(attachmentReference.getName())
+            .getContentInputStream(wikiContext); ByteArrayOutputStream baos = new ByteArrayOutputStream())
+        {
+            OfficeManager manager = ExternalOfficeManager.builder().portNumbers(8100).build();
```

Comment:
The port number should be taken from the xwiki configuration, since this can be changes by the user